nxp: k6x: move clk divider options to device tree

aunsbjerg opened the pull request:
With inspiration from #26610, I've started moving the clock divider settings for the SIM module from the SoC's kconfig to device tree. The defaults are now set in `nxp_k6x.dtsi` and can be overwritten by boards as needed.

Commit:
soc: nxp: k6x: move clk divider options to device tree

Use kinetis SIM clock divider options set in device tree
instead of kconfig

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@prevas.dk>

Commit:
boards: arm: ip_k66f: set clk dividers in device tree

Set kinetis SIM clock divider options in device tree
instead of in SoC's kconfig

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@prevas.dk>

henrikbrixandersen commented:
@aunsbjerg You need to fix the compliance issues with the missing bodies in the Git commit messages.

galak reviewed (changes requested):
So I think this should be more inline with clock devicetree bindings:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/clock/fixed-factor-clock.yaml

So something like:

```
core_clk: core_clk {
        #clock-cells = <0>;
        compatible = "fixed-factor-clock";
        clocks = <&mcgoutclk>;
        clock-div = <1>;
};
```

Kludentwo commented on soc/arm/nxp_kinetis/k6x/soc.h:
```diff
@@ -24,8 +24,12 @@
+
 #define BUSCLK_DEFAULT_IOSC_HZ (SYSCLK_DEFAULT_IOSC_HZ / \
-				CONFIG_K6X_BUS_CLOCK_DIVIDER)
+				CLOCK_DIVIDER(bus_clk_divider))
```
CONFIG_K6X_BUS_CLOCK_DIVIDER was 2 but using the macro makes it 1?

aunsbjerg commented:
> So I think this should be more inline with clock devicetree bindings:
> 
> https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/clock/fixed-factor-clock.yaml
> 
> So something like:
> 
> ```
> core_clk: core_clk {
>         #clock-cells = <0>;
>         compatible = "fixed-factor-clock";
>         clocks = <&mcgoutclk>;
>         clock-div = <1>;
> };
> ```

I added a `fixed-factor-clock` binding as suggested and end up with something like this:

```
// nxp_k6x.dtsi
sim: sim@40047000 {
	compatible = "nxp,kinetis-sim";
	reg = <0x40047000 0x1060>;
	label = "SIM";
	#clock-cells = <3>;

	core_clk: core_clk {
		compatible = "fixed-factor-clock";
		clocks = <&mcg KINETIS_MCG_FIXED_FREQ_CLK>; // need to add proper clock name
		clock-div = <1>;
		#clock-cells = <0>;
	}; 
};
```
And:
```
// k6x soc.c
#define CLOCK_DIVIDER(clk) \
	DT_PROP_OR(DT_NODELABEL(clk), clock_div, 1) - 1

static const sim_clock_config_t simConfig = {
	.pllFllSel = DT_PROP(DT_INST(0, nxp_kinetis_sim), pllfll_select),
	.er32kSrc = DT_PROP(DT_INST(0, nxp_kinetis_sim), er32k_select),
	.clkdiv1 = SIM_CLKDIV1_OUTDIV1(CLOCK_DIVIDER(core_clk)) |
		   SIM_CLKDIV1_OUTDIV2(CLOCK_DIVIDER(bus_clk)) |
		   SIM_CLKDIV1_OUTDIV3(CLOCK_DIVIDER(flexbus_clk)) |
		   SIM_CLKDIV1_OUTDIV4(CLOCK_DIVIDER(flash_clk)),
};
```

However, I'm not sure if this is what you are suggesting or if this is the way to go. This differs quite a bit from the `nxp,kinetis-scg.yaml` binding + driver for the `ke1xf` SoC that has most (if not all) clock settings available in device tree. Thoughts?

galak commented:
> However, I'm not sure if this is what you are suggesting or if this is the way to go. This differs quite a bit from the `nxp,kinetis-scg.yaml` binding + driver for the `ke1xf` SoC that has most (if not all) clock settings available in device tree. Thoughts?

What you've suggested is exactly inline with what I'm thinking and asking for.  The support for the `ke1xf` should eventually be changed over to look like what you've done.  Its a historic example that we should try and avoid propagating and fix as we can.

The reason is over time we'll want to have proper clock tree's in devicetree to handle power management correctly.  At the level of support we've had historically this wasn't as large of a concern.  However as we keep improving things this is the direction we should head.

Commit:
drivers: clock_control: mcux_mcg: add MCGOUTCLK define

Add MCGOUTCLK define to kinetis_mcg.h to make it possible to
use \`<&mcg KINETIS_MCG_OUT_CLK>\` in device tree.

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@prevas.dk>

Commit:
dts: bindings: clock: add fixed-factor-clock binding

Allows for adding in fixed factor clock that are derived
from a parent clock.

Based on:
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/clock/fixed-factor-clock.yaml

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@prevas.dk>

aunsbjerg commented:
> > However, I'm not sure if this is what you are suggesting or if this is the way to go. This differs quite a bit from the `nxp,kinetis-scg.yaml` binding + driver for the `ke1xf` SoC that has most (if not all) clock settings available in device tree. Thoughts?
> 
> What you've suggested is exactly inline with what I'm thinking and asking for. The support for the `ke1xf` should eventually be changed over to look like what you've done. Its a historic example that we should try and avoid propagating and fix as we can.
> 
> The reason is over time we'll want to have proper clock tree's in devicetree to handle power management correctly. At the level of support we've had historically this wasn't as large of a concern. However as we keep improving things this is the direction we should head.

Super, that makes sense. The more I looked at the existing code, the more confused I got.
I've pushed some changes that I believe implements what you are suggesting.

galak commented on dts/bindings/clock/fixed-rate-clock.yaml:
```diff
@@ -0,0 +1,27 @@
+
+properties:
+    label:
+      type: string
```
Remove `label` its not needed here.

galak reviewed (commented):
LGTM.  One change on dropping `label` from the binding.

Can you look at making this change to the other relevant Kinetis SoCs:

```
soc/arm/nxp_kinetis/k2x/Kconfig.soc:config K22_FLASH_CLOCK_DIVIDER
soc/arm/nxp_kinetis/k2x/soc.c:             SIM_CLKDIV1_OUTDIV4(CONFIG_K22_FLASH_CLOCK_DIVIDER - 1),
soc/arm/nxp_kinetis/k6x/Kconfig.soc:config K6X_FLASH_CLOCK_DIVIDER
soc/arm/nxp_kinetis/k6x/soc.c:             SIM_CLKDIV1_OUTDIV4(CONFIG_K6X_FLASH_CLOCK_DIVIDER - 1),
soc/arm/nxp_kinetis/k8x/Kconfig.soc:config K8X_FLASH_CLOCK_DIVIDER
soc/arm/nxp_kinetis/k8x/soc.c:             SIM_CLKDIV1_OUTDIV4(CONFIG_K8X_FLASH_CLOCK_DIVIDER - 1),
soc/arm/nxp_kinetis/kv5x/Kconfig.soc:config KV5X_FLASH_CLOCK_DIVIDER
soc/arm/nxp_kinetis/kv5x/soc.c:            SIM_CLKDIV1_OUTDIV4(CONFIG_KV5X_FLASH_CLOCK_DIVIDER - 1),
soc/arm/nxp_kinetis/kwx/Kconfig.soc:config KW2XD_FLASH_CLOCK_DIVIDER
soc/arm/nxp_kinetis/kwx/soc_kw2xd.c:               SIM_CLKDIV1_OUTDIV4(CONFIG_KW2XD_FLASH_CLOCK_DIVIDER - 1),
```

Commit:
soc: nxp: kwx: move clk divider options to device tree

Use kinetis SIM clock divider options set in device tree
instead of kconfig.

Both the kw40z and kw41z device tree originally used an
undefined "nxp,kw41z-mcg" binding for the MCG node.
This has been replaced with the general "nxp,kinetis-mcg"
binding instead.

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@prevas.dk>

Commit:
soc: nxp: kv5x: move clk divider options to device tree

Use kinetis SIM clock divider options set in device tree
instead of kconfig.

The kv5x device tree originally used the undefined
"nxp,kv58-mcg" binding for the MCG node. This has been
replaced by the general "nxp,kinetis-mcg" binding.

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@prevas.dk>

Commit:
soc: nxp: k8x: move clk divider options to device tree

Use kinetis SIM clock divider options set in device tree
instead of kconfig

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@prevas.dk>

Commit:
soc: arm: nxp: remove unused clock defines

SYSCLK_DEFAULT_IOSC_HZ and BUSCLK_DEFAULT_IOSC_HZ are
not used anywhere in the tree and can be removed

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@prevas.dk>

Commit:
soc: nxp: k2x: move clk divider options to device tree

Use kinetis SIM clock divider options set in device tree
instead of kconfig

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@prevas.dk>

Commit:
soc: nxp: kl2x: move clk divider options to device tree

Use kinetis SIM clock divider options set in device tree
instead of hardcoded values.

The kl25z device tree did not previously define a MCG node.
This has now been added with the general "nxp,kinetis-mcg"
binding.

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@prevas.dk>

aunsbjerg commented:
> LGTM. One change on dropping `label` from the binding.
> 
> Can you look at making this change to the other relevant Kinetis SoCs:
> 
> ```
> soc/arm/nxp_kinetis/k2x/Kconfig.soc:config K22_FLASH_CLOCK_DIVIDER
> soc/arm/nxp_kinetis/k2x/soc.c:             SIM_CLKDIV1_OUTDIV4(CONFIG_K22_FLASH_CLOCK_DIVIDER - 1),
> soc/arm/nxp_kinetis/k6x/Kconfig.soc:config K6X_FLASH_CLOCK_DIVIDER
> soc/arm/nxp_kinetis/k6x/soc.c:             SIM_CLKDIV1_OUTDIV4(CONFIG_K6X_FLASH_CLOCK_DIVIDER - 1),
> soc/arm/nxp_kinetis/k8x/Kconfig.soc:config K8X_FLASH_CLOCK_DIVIDER
> soc/arm/nxp_kinetis/k8x/soc.c:             SIM_CLKDIV1_OUTDIV4(CONFIG_K8X_FLASH_CLOCK_DIVIDER - 1),
> soc/arm/nxp_kinetis/kv5x/Kconfig.soc:config KV5X_FLASH_CLOCK_DIVIDER
> soc/arm/nxp_kinetis/kv5x/soc.c:            SIM_CLKDIV1_OUTDIV4(CONFIG_KV5X_FLASH_CLOCK_DIVIDER - 1),
> soc/arm/nxp_kinetis/kwx/Kconfig.soc:config KW2XD_FLASH_CLOCK_DIVIDER
> soc/arm/nxp_kinetis/kwx/soc_kw2xd.c:               SIM_CLKDIV1_OUTDIV4(CONFIG_KW2XD_FLASH_CLOCK_DIVIDER - 1),
> ```

Done. Let me know if you prefer to have some of the commits squashed.
Also, b82b656 removed some unused defines, which is not strictly part of this PR. 
The blinky sample still works on frdm_k64f and frdm_k22f boards - don't have boards to test the other SoC changes.

aunsbjerg commented:
Buildkite errors should be fixed now.

The issues were caused by a missing mcg node in the kl25z board and the kw40z, kw41z and kv5x device tree using undefined bindings for the mcg instead of using "nxp,kinetis-mcg".

MaureenHelm reviewed (approved):
This looks great, thank you @aunsbjerg ! We've needed to start moving clocks to dts for a while now.

galak commented:
@aunsbjerg any reason for this to still be labeled DNM?

aunsbjerg commented:
> @aunsbjerg any reason for this to still be labeled DNM?

No, just me not paying attention. Fixed.

henrikbrixandersen reviewed (approved):
Looks good, thank you.
